### ISSUE TITLE
Getting/checking properties of fetched algorithms

### ISSUE THREAD

**Issue by beldmit** [openssl/openssl#17470]
Is there any API designed for obtaining algorithm properties for the fetched algorithm or checking whether a particular property is set?

If no, will this API be appropriated for including in 3.x series?

**Comment by t8m** [comment]
No, there is no API for querying properties. IMO we could/should add such API (in master).

**Comment by paulidale** [comment]
An API to get an algorithm's property string wouldn't be difficult.  Exposing the internal functions to parse and query these strings would be more problematic but not impossible.

What's the use case for needing to query these?

**Comment by beldmit** [comment]
Not sure I can describe the use case here, sorry.

**Comment by beldmit** [comment]
We would like to add some extra properties to algorithms and check their presence but not set them as default ones when fetching.

**Comment by paulidale** [comment]
I'm not sure I understand fully.

Why not fetch with the extra properties and see if something is returned?

**Comment by levitte** [comment]
Who's "we", @beldmit?

There is an internal API, which the encoder and decoder library code uses:

https://github.com/openssl/openssl/blob/a10a576090022e583a06271ceced8e38dd509657/include/internal/property.h#L42-L49

This is currently only available from within libcrypto

**Comment by mattcaswell** [comment]
I guess something like:

````
EVP_CIPHER *ciph = EVP_CIPHER_fetch(NULL, "MYALG", NULL);

if (ciph has property "myproperty" set) {
    /* do one thing with ciph */
} else {
   /* do another thing with ciph */
}
````

An interesting idea!

**Comment by mattcaswell** [comment]
E.g. it could be used to declare things like "I support in-place encryption"

**Comment by paulidale** [comment]
This can be done currently in a slightly roundabout fashion:

```c
EVP_CIPHER *ciph1 = EVP_CIPHER_fetch(NULL, "MYALG", "myproperty");
EVP_CIPHER *ciph2 = EVP_CIPHER_fetch(NULL, "MYALG", NULL);

if (ciph1 != NULL) {
    /* do one thing with ciph1 */
} else {
   /* do another thing with ciph2 */
}
```

**Comment by mattcaswell** [comment]
> This can be done currently in a slightly roundabout fashion:

Yes that would work....but being able to query the properties directly is neater and nicer.

**Comment by paulidale** [comment]
Slower....

**Comment by mattcaswell** [comment]
> Slower....

You mean doing 2 fetches is slower? Yes.

**Comment by paulidale** [comment]
No, querying to see which you've got is slower....
The fetches can be cached.  I guess the query could too.

**Comment by beldmit** [comment]
> I guess something like:
> 
> ```
> EVP_CIPHER *ciph = EVP_CIPHER_fetch(NULL, "MYALG", NULL);
> 
> if (ciph has property "myproperty" set) {
>     /* do one thing with ciph */
> } else {
>    /* do another thing with ciph */
> }
> ```
> 
> An interesting idea!

yes, smth like that.

**Comment by beldmit** [comment]
> Who's "we", @beldmit?

Sorry. s/we/users/

**Comment by beldmit** [comment]
FIPS 140-3 requires some sort of indicators, either implicit (algorithm is fetched and everything is OK) or explicit (algorithm is fetched, we check if it is OK, and use it). Having an access to properties will allow to do some tricks - e.g. allowing using sha1 in HMAC only but not in naked form.

**Comment by paulidale** [comment]
Something like this achieves the desired result:
```c
hmac = EVP_MAC_fetch(NULL, "hmac", "fips3");
ctx = EVP_MAC_CTX_new(hmac);
params[0] = OSSL_PARAM_construct_utf8_string("digest", "sha1", 0);
params[1] = OSSL_PARAM_construct_utf8_string("properties", "fips,-fips3", 0);
...
EVP_MAC_init(ctx, key, keylen, params);
```

We will need a more general solution for fetching FIPS algorithms where they are only valid for some uses.  Or we abrogate this to the user via the security policy.

**Comment by beldmit** [comment]
I agree that it's a possible workaround, but a more general solution and API for explicit options checks seems to be worth implementing.
